### ISSUE TITLE
Problem using prettier as a git hook.

### ISSUE THREAD

**Issue by migrachev** [prettier/prettier#2978]
**[Explaning my problem]**
Hello I am trying to use prettier as pre-commit git hook (along with WebStorm).
I had the problem that even after I commit my changes, my files are marked as changed and **_git status_** and **_git diff_** output something strange. I spent some time investigating the problem and I reproduce it directly from the command line, excluding WebStorm from the problem. Since the problem occurs only when prettier prettifies my files my conclusion was that the problem must be in prettier not git.

**[If you are in a hurry]**
Just go to **My Simple Example**

**[The actual problem]**
When I commit a file that is considered **_ugly_** by prettier, the git hook that i have defined invoke prettier on the file, makes it **_pretty_** and then git performs the commit. At that point my expectations are that there will be no changes in my working directory, however git does not think so.

**[Why do I think the problem comes from prettier and not git]**
- If the changed file is considered **_pretty_** by prettier, it seems that prettier does not process it and the problem does not reproduce.
- According to the internet changing the source code in pre-commit git hooks seems to be a practice and it should work fine.

**[My Simple Example]**
- Note that I am using the prettier pre-commit hook from https://github.com/prettier/prettier I only added some additional prettier options that I need.
https://s1.postimg.org/20gmod0x73/With_Git_Attributes.png

**Comment by azz** [comment]
I am able to reproduce this. The following line:

```bash
echo "$jsfiles" | xargs git add
```

Is failing to re-stage the `example.js` file. 

I found [this StackOverflow](https://stackoverflow.com/a/3971851/1280997) question and it seems that there are issues with `git add` during pre-commit on Windows.

**Comment by migrachev** [comment]
Actually it seems it is not failing to re-stage the files but git indexes are getting messy.
Haven't test the issue on Linux, can't say if this is only Windows problem, but here is what I found today:
https://stackoverflow.com/questions/41401296/understanding-git-commit-only-and-pre-commit-hooks

It seems that apart from pre-commit hook we need a post-commit one to update git indexes.
For me this works fine at the moment both from the command line and from WebStorm.

I am not sure if that is the correct solution, because I seem to lack git understanding on the topic, but it we all agree this is the solution, I recommend this post-commit hook to become part of the documentation @ https://github.com/prettier/prettier next to pre-commit one.

Let me know what do you think.

**Comment by duailibe** [comment]
@migrachev Would you add a small explanation to our docs (maybe with a link to the StackOverflow question) ?

**Comment by migrachev** [comment]
@duailibe Sorry that I see my notification that late. I was going to fork the repo and make a pull request, but i notised that the documentation was moved out of git into https://prettier.io/docs so I don't know if I am able to contribute to the documentation part at the moment.

However the solution of the problem that I describe above still works fine for me and my team, so I consider it stable. Here is what we have done:
1. We are using the "bash script option" from the documentaion - https://prettier.io/docs/en/precommit.html
2. In addition to the pre-commit hook, we have a post commit hook to correct git indexes. It looks like this:
````js
#!/bin/sh
git update-index -g
exit 0
````
Feel free to add it to the docs if you consider it worth.

**Comment by j-f1** [comment]
@migrachev The docs are in the `docs/` directory in the repo.

**Comment by azz** [comment]
Specifically: https://github.com/prettier/prettier/blob/master/docs/precommit.md

**Comment by lsuarez-tiempo-dev** [comment]
I understand the adding the git bash script entails changing the execution permissions with something like chmod 777. not something I'd like  to ask everyone one on my team to do.
I was thinking on doing this via husky. 
Is there any reason why this would be a bad idea?

``` 
husky": {
    "hooks": {
      "pre-commit": "pretty-quick --staged",
      "post-commmit":"git update-index -g"
    }
  },
```

**Comment by j-f1** [comment]
Nope, that’s actually the method suggested as [option 2](https://prettier.io/docs/en/precommit.html#option-2-pretty-quick-https-githubcom-azz-pretty-quick) in the docs.

**Comment by j-f1** [comment]
Can we close this? The PR referencing the issue seems to have been merged.

**Comment by lsuarez-tiempo-dev** [comment]
One last thing.
Option 2 doesn't mention the post commit hook for updating the git index. I
haven't tested it yet.
Testing right now..

**Comment by lsuarez-tiempo-dev** [comment]
works great !

**Comment by migrachev** [comment]
@j-f1 yes I believe we can close the issue, the problem on it have been solved.

**Comment by JonasGroeger** [comment]
According to https://youtrack.jetbrains.com/issue/IDEA-135454 this issue will be fixed in IntelliJ IDEA 2018.3
